mqtt-level-store.js

```diff
@@ -19,38 +19,112 @@ function Store (options) {
   }
 }
 
+const zeroPadding = (function () {
+  // width and padding calculates only once
+  const width = parseInt(Number.MAX_SAFE_INTEGER).toString(10).length
+  const padding = new Array(width + 1).join('0')
+  return function (num) {
+    return (padding + num).slice(-width)
+  }
+})()
+
 Store.prototype.put = function (packet, cb) {
-  this._level.put('' + packet.messageId, packet, this._levelOpts, cb)
+  var date = zeroPadding(Date.now())
+  if (this._prevDate === date) {
+    ++this._sameDateCount
+  } else {
+    this._sameDateCount = 0
+    this._prevDate = date
+  }
+  date += zeroPadding(this._sameDateCount)
+
+  var that = this
+  this._level.get(
+    'packets~' + packet.messageId,
+    this._levelOpts,
+    function (err, _date) {
+      if (err) {
+        if (err.notFound) {
+          var cmd = [
+            {type: 'put', key: 'packets~' + packet.messageId, value: date},
+            {type: 'put', key: 'packet-by-date~' + date + '~' + packet.messageId, value: packet}
+          ]
+          that._level.batch(cmd, that._levelOpts, cb)
+        } else {
+          cb(err)
+        }
+      } else {
```

Comment:
can you just return early from the previous block instead of adding this else?

Reply:
Do you mean as follows?

```cpp
      if (err) {
        if (!err.notFound) return cb(err)
        var cmd = [
          {type: 'put', key: 'packets~' + packet.messageId, value: date},
          {type: 'put', key: 'packet-by-date~' + date + '~' + packet.messageId, value: packet}
        ]
        that._level.batch(cmd, that._levelOpts, cb)
      } else {
```

Or

```cpp
        if (!err.notFound) return cb(err)
```

is

```cpp
        if (!err.notFound) {
          cb(err)
          return this
        }
```

?

Or something else?

Reply:
I understand the first one is appropriate. Similar to https://github.com/mcollina/mqtt-level-store/pull/11#discussion_r212822705